File: MotoActions/src/org/lineageos/settings/device/doze/StowSensor.java
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.doze;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.util.Log;

import org.lineageos.settings.device.MotoActionsSettings;
import org.lineageos.settings.device.SensorAction;
import org.lineageos.settings.device.SensorHelper;

public class StowSensor implements ScreenStateNotifier {
    private static final String TAG = "MotoActions-StowSensor";

    // Maximum time for the hand to cover the sensor: 1s
    private static final long HANDWAVE_MAX_DELTA_NS = 1000L * 1000 * 1000;

    // Minimum time until the device is considered to have been in the pocket: 5s
    private static final long POCKET_MIN_DELTA_NS = 5000L * 1000 * 1000;

    private final MotoActionsSettings mMotoActionsSettings;
    private final SensorHelper mSensorHelper;
    private final SensorAction mSensorAction;
    private final Sensor mStowSensor;

    private boolean mEnabled;
    private boolean mLastStowed;
    private long mLastStowedTime;

    public StowSensor(MotoActionsSettings MotoActionsSettings, SensorHelper sensorHelper,
                      SensorAction action) {
        mMotoActionsSettings = MotoActionsSettings;
        mSensorHelper = sensorHelper;
        mSensorAction = action;

        mStowSensor = sensorHelper.getStowSensor();
    }

    @Override
    public void screenTurnedOn() {
        if (mEnabled) {
            Log.d(TAG, "Disabling");
            mSensorHelper.unregisterListener(mStowListener);
            mEnabled = false;
        }
    }

    @Override
    public void screenTurnedOff() {
        if ((mMotoActionsSettings.isPocketGestureEnabled()
                || mMotoActionsSettings.isIrWakeupEnabled())
                && !mEnabled) {
            Log.d(TAG, "Enabling");
            mSensorHelper.registerListener(mStowSensor, mStowListener);
            mEnabled = true;
        }
    }

    private final SensorEventListener mStowListener = new SensorEventListener() {
        @Override
        public synchronized void onSensorChanged(SensorEvent event) {
            boolean thisStowed = (event.values[0] != 0);
            if (thisStowed) {
                mLastStowedTime = event.timestamp;
            } else if (mLastStowed) {
                if (shouldPulse(event.timestamp)) {
                    mSensorAction.action();
                }
            }
            mLastStowed = thisStowed;
            Log.d(TAG, "event: " + thisStowed);
        }

        private boolean shouldPulse(long timestamp) {
            long delta = timestamp - mLastStowedTime;

            boolean irWakeupEnabled = mMotoActionsSettings.isIrWakeupEnabled();
            boolean pocketGestureEnabled = mMotoActionsSettings.isPocketGestureEnabled();

           if (irWakeupEnabled && pocketGestureEnabled) {
                return true;
            } else if (irWakeupEnabled) {
                return delta < HANDWAVE_MAX_DELTA_NS;
            } else if (pocketGestureEnabled) {
                return delta >= POCKET_MIN_DELTA_NS;
            }
            return false;
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int accuracy) {
        }
    };
}

File: packages/MotoActions/src/org/lineageos/settings/device/doze/StowSensor.kt
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.doze

import android.content.Context
import android.content.SharedPreferences
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.util.Log
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_IR_WAKEUP_KEY
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_POCKET_KEY
import org.lineageos.settings.device.SensorHelper
import org.lineageos.settings.device.SensorHelper.Companion.SENSOR_TYPE_MMI_STOW

class StowSensor(
    context: Context,
    private val sharedPreferences: SharedPreferences,
    private val sensorHelper: SensorHelper,
) : ScreenStateNotifier {

    private val dozePulseAction: DozePulseAction = DozePulseAction(context)
    private val stowSensor: Sensor = sensorHelper.getSensor(SENSOR_TYPE_MMI_STOW)!!

    private var lastStowed = false
    private var lastStowedTime = 0L

    override fun onScreenStateChanged(screenOn: Boolean) {
        if (screenOn) {
            if (enabled) {
                Log.d(TAG, "Disabling")
                sensorHelper.unregisterListener(stowListener)
                enabled = false
            }
        } else {
            if (
                (sharedPreferences.getBoolean(GESTURE_POCKET_KEY, true) ||
                    sharedPreferences.getBoolean(GESTURE_IR_WAKEUP_KEY, true)) && !enabled
            ) {
                Log.d(TAG, "Enabling")
                sensorHelper.registerListener(stowSensor, stowListener)
                enabled = true
            }
        }
        dozePulseAction.onStateChanged(false)
    }

    private val stowListener =
        object : SensorEventListener {
            @Synchronized
            override fun onSensorChanged(event: SensorEvent) {
                val thisStowed = event.values[0] != 0f
                if (thisStowed) {
                    lastStowedTime = event.timestamp
                } else if (lastStowed && shouldPulse(event.timestamp)) {
                    dozePulseAction.onStartPulse()
                }
                lastStowed = thisStowed
                Log.d(TAG, "event: $thisStowed")
            }

            private fun shouldPulse(timestamp: Long): Boolean {
                val delta = timestamp - lastStowedTime

                val irWakeupEnabled = sharedPreferences.getBoolean(GESTURE_IR_WAKEUP_KEY, true)
                val pocketGestureEnabled = sharedPreferences.getBoolean(GESTURE_POCKET_KEY, true)

                return when {
                    irWakeupEnabled && pocketGestureEnabled -> true
                    irWakeupEnabled -> delta < HANDWAVE_MAX_DELTA_NS
                    pocketGestureEnabled -> delta >= POCKET_MIN_DELTA_NS
                    else -> false
                }
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }

    companion object {
        private const val TAG = "MotoActions-StowSensor"
        private const val HANDWAVE_MAX_DELTA_NS = 1000L * 1000 * 1000
        private const val POCKET_MIN_DELTA_NS = 5000L * 1000 * 1000
    }
}
